mono/metadata/superluminal.c

```diff
@@ -0,0 +1,330 @@
+}
+
+static void
+on_enumerate_assembly (MonoAssembly *assembly, void *user_data)
+{
+	struct JITEnumerationData *enumerationData = (struct JITEnumerationData *)user_data;
+	enumerationData->mNumAssemblies++;
+
+	MonoImage *image = mono_assembly_get_image (assembly);
+	image_event (image, TRUE, FALSE);
+}
+
+static void
+on_enumerate_jit_method (MonoDomain *domain, MonoMethod *method, MonoJitInfo *jinfo, void *user_data)
+{
+	struct JITEnumerationData *enumerationData = (struct JITEnumerationData *)user_data;
+	enumerationData->mNumMethods++;
+
+	method_load (domain, method, jinfo, TRUE);
+}
+
+static void
+on_enumerate_domain (MonoDomain *domain, void *user_data)
+{
+	struct JITEnumerationData *enumerationData = (struct JITEnumerationData *)user_data;
+	enumerationData->mNumDomains++;
+
+	// Iterate through each assembly
+	mono_domain_assembly_foreach (domain, on_enumerate_assembly, enumerationData);
+
+	// Iterate through each JIT'ed method
+	mono_domain_jit_foreach (domain, on_enumerate_jit_method, enumerationData);
+}
+
+static void
+on_attach ()
+{
+	SUPERLUMINAL_LOG ("Enumerating JIT data...");
+
+	struct JITEnumerationData enumerationData;
+	memset (&enumerationData, 0, sizeof (enumerationData));
+
+	// Iterate through each domain
+	mono_domain_foreach (on_enumerate_domain, &enumerationData);
+
+	SUPERLUMINAL_LOG_ARGS ("Finished enumerating JIT data. Found %d domains, %d assemblies, %d methods", enumerationData.mNumDomains, enumerationData.mNumAssemblies, enumerationData.mNumMethods);
+}
+
+// This callback is called by the ETW system when tracing is started / stopped. We use it to enumerate & output information about JIT compilation that happened *before* tracing started.
+DECLSPEC_NOINLINE __inline VOID __stdcall Private_EventControlCallback (_In_ LPCGUID SourceId, _In_ ULONG ControlCode, _In_ UCHAR Level, _In_ ULONGLONG MatchAnyKeyword, _In_ ULONGLONG MatchAllKeyword, _In_opt_ PEVENT_FILTER_DESCRIPTOR FilterData, _Inout_opt_ PVOID CallbackContext)
+{
+	SUPERLUMINAL_LOG_ARGS ("EventControlCallback (%d)", ControlCode);
+
+	switch (ControlCode) {
+	case EVENT_CONTROL_CODE_ENABLE_PROVIDER: {
+		gboolean isRegular = (CallbackContext == &MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_Context);
+		gboolean isRundown = (CallbackContext == &MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_Context);
+		SUPERLUMINAL_LOG_ARGS("EventControlCallback -- IsInitialized: %s, IsRuntime: %s, IsRundown: %s", is_initialized ? "true" : "false", isRegular ? "true" : "false", isRundown ? "true" : "false");
+		if (is_initialized && CallbackContext == &MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_Context) {
+			on_attach ();
+		}
+	} break;
+	default:
+		break;
+	};
+}
+
+/* the entry point */
+MONO_API void
+mono_profiler_init_superluminal (const char *desc)
+{
+	SUPERLUMINAL_LOG ("Initializing Plugin");
+
+	EventRegisterMicrosoft_Windows_DotNETRuntime ();
+	EventRegisterMicrosoft_Windows_DotNETRuntimeRundown ();
```

Comment:
Is there any way we could just have this be called by mono init on startup without having Unity manually call this? All these methods are very cheap, and return immediately if ETW recording is not enabled. CoreCLR has this enabled at all times.

Reply:
Yes, this will be enabled by default in the associate Unity repo pull request.

Reply:
Why not just call it from `mono_init`, so that embedders don't need to do anything special? This would make it functional when using Mono to run tools too (IDK if we still do it), and in non-development players as well.

Reply:
I don't think there's much use in moving the register calls outside of the plugin. They don't do anything other than registration of the ETW providers, so even if this was done more centrally, you'd still need the profiler plugin to actually emit the ETW events based on the profiler callbacks.

Reply:
What profiler plugin are you talking about? I tested this and I see .NET ETW events get emitted after I call this function in an empty Unity project.

Reply:
Sorry, I think I misunderstood - I thought you were talking about just moving the `EventRegister*` calls somewhere else, but I guess you were talking about moving the call to `mono_profiler_init_superluminal` somewhere central. My bad :-)

Reply:
Yup! I just think that Mono should be registering these events unconditionally.

Reply:
> This would make it functional when using Mono to run tools too (IDK if we still do it)

Most of the tools run from the Unity Editor are using .NET now, I don't think that we use the Mono we ship for them any more.

Reply:
> in non-development players as well

I had not thought about this aspect. Based on the Unity repo change (sorry @rovarma, those are in an internal repo). I think this will be enabled even for non-development players.